### DISCUSSION TITLE
How to make exceptions to --private and --private-home?

### DISCUSSION THREAD

**Discussion by diamond-lizard** [netblue30/firejail#5225] [Q&A] [answered]
I start firefox under firejail with:

    firejail --private --apparmor --private-home=.mozilla firefox -no-remote

Because of this any changes that firefox makes to the sandbox (such as any changes to my bookmarks or history, and anything downloaded to my ```~/Downloads``` folder) are lost when firefox exits.

Is there any way to make exceptions to ```--private``` and ```--private-home``` such that these changes are preserved but any other changes are discarded when firefox exits?

**Comment by rusty-snake** [comment]
> Is there any way to make exceptions to --private and --private-home such that these changes are preserved but any other changes are discarded when firefox exits?

No.

For `--private` this wouldn't even make sense / would be good.

Why do you combine `--private` and `--private-home`? Does `--private` make any differences here?

> **Note**: You can use directories you side of your home directory or FILE TRANSFER to keep downloaded files.

**Reply by diamond-lizard** [reply]
> Why do you combine --private and --private-home? Does --private make any differences here?

If I use only ```--private``` then Firefox starts up as if I started it for the first time, without any of my extensions or any configuration settings that I made to it because it does not have access to the contents of my firefox profile in ```~/.mozilla```

When I added ```--private-home=.mozilla``` then it started up with my extensions and configuration settings as I would normally expect.  The only problem is that any additions to my bookmarks and any downloads I make disappear when firefox exits.

**Comment by ghost** [comment]
On top of @rusty-snake's comments about --private and --private-home being non-combinable, the `--apparmor` flag is superfluous here too, as it is already enabled in [firefox-common.profile](https://github.com/netblue30/firejail/blob/master/etc/profile-a-l/firefox-common.profile#L35).

OT: @rusty-snake @smitsohu @netblue30 This incompatibility is mentioned in man firejail (--private-home stanza). Perhaps it would be wise to reflect this in the codebase and catch using them together early-on? Not sure if we actually want this kind of 'do-not-shoot-yourself-in-the-foot' prevention.

> Because of this any changes that firefox makes to the sandbox (such as any changes to my bookmarks or history, and anything downloaded to my ~/Downloads folder) are lost when firefox exits.

That's indeed the result of using `--private-home=.mozilla`. Maybe we could be more helpful if you tell us what it is you're trying to achieve exactly.

**Reply by diamond-lizard** [reply]
> Maybe we could be more helpful if you tell us what it is you're trying to achieve exactly.

I want to run firefox under firejail, and when firefox exits I want to keep the files that I download, any changes I make to my bookmarks, and my browsing history, while discarding any other changes firefox might happen to make (my concern being potential malware that firefox might be vulnerable to during my web browsing).

**Reply by rusty-snake** [reply]
> This incompatibility is mentioned in man firejail (--private-home stanza). Perhaps it would be wise to reflect this in the codebase and catch using them together early-on? Not sure if we actually want this kind of 'do-not-shoot-yourself-in-the-foot' prevention.

:+1:

I would it make like `--profile=foo --noprofile` already looks.

**Comment by rusty-snake** [comment, marked as answer]
> I want to run firefox under firejail, and when firefox exits I want to keep the files that I download, any changes I make to my bookmarks, and my browsing history, while discarding any other changes firefox might happen to make (my concern being potential malware that firefox might be vulnerable to during my web browsing).

This is exactly¹ what our firefox.profile already does. Look at it and the documentation.
- `--private` is the wrong technology for you as it means start without anything (and `--private-home` makes this flag ignored anyway)
- `--private-home` it the wrong technology for you either as it means discard everything.

¹ not strict but to the degree I care.

---

General advises: Avoid XY Problems when asking questions. So you question should have been like "How do I create a sandbox that discards every change except for Downloads/Bookmarks/Updates?" And not "I think I can combine --private and --private-home to create this. But now I need to ask how to make exceptions to it".
